ProdPerfect/prodperfect-keen-tracking.js
# Implements throttling functionality

# Description

This feature uses a hardcoded "blacklist" of `PROJECT_ID`s to selectively
throttle what events get sent back to the ProdPerfect servers.

# Pseudocode
```javascript
IF (blacklist.contains(PROJECT_ID) {
  IF (should_this_session_uuid_be_sampled?()) {
     return continue_as_normal();
  } ELSE {
     return short_circuit_setup_logic_and_do_nothing_further();
  }
}
```

## Further details
To determine whether or not a specific session should be sampled:

Calculate a MD5 hash from the `session_uuid`, and use the last 8 hex valued
characters of the hash to get a number (e.g. `0x499602d2` becomes the int `1234567890`)
Compare this number to the highest possible 8 character hex value (`0xffffffff` or `4294967295`) as a ratio:
```
1234567890/4294967295 = 0.2874..
```

...or approx **29%**

If we set our throttle value for this project ID to 50%, then the 29% outcome value would mean that we
**should** sample this session.

If we set our throttle value to something lower like 5%, then the 29% value would mean we would **not**
sample this value.

Since the `session_uuid` is a UUID and we use hashing, we can rely on the numerical values we extract
to be more-or-less normally distributed, and can use it as a means to decide whether or not the session
should be sampled.

# How to verify this feature

1) Check out the [unit tests]() to get a feel for how this module is used.
2) Add the [demo project id]() to the [blacklist] on your local machine
3) Run `npm run start` in your terminal

If you set the throttle value high enough (e.g to `1.00`), you will notice that the tracker does **not** send events as expected (because your session is not being sampled).  Similarly if you set your throttle value low enough, odds are your events will be sent.  Play around with the value.

There is also [a unit test]() that checks whether or not to sample against 10 000 generated UUIDs, to ensure that statistically the algo performs as advertised.  You could do this in your console as well if you feel inclined.


Thanks for reading!
